core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/hashJoin/IncrementalJoinOpExec.scala

```diff
@@ -0,0 +1,124 @@
+package edu.uci.ics.texera.workflow.operators.hashJoin
+
+import edu.uci.ics.amber.engine.architecture.worker.PauseManager
+import edu.uci.ics.amber.engine.common.InputExhausted
+import edu.uci.ics.amber.engine.common.rpc.AsyncRPCClient
+import edu.uci.ics.texera.workflow.common.operators.OperatorExecutor
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import edu.uci.ics.texera.workflow.common.tuple.Tuple.BuilderV2
+import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, OperatorSchemaInfo, Schema}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+class IncrementalJoinOpExec[K](
+    val buildAttributeName: String,
+    val probeAttributeName: String,
+    val operatorSchemaInfo: OperatorSchemaInfo
+) extends OperatorExecutor {
+
+  val leftSchema: Schema = operatorSchemaInfo.inputSchemas(0)
+  val rightSchema: Schema = operatorSchemaInfo.inputSchemas(1)
+
+  val leftTuples = new mutable.HashMap[K, (ArrayBuffer[Tuple], Boolean)]()
+  val rightTuples = new mutable.HashMap[K, (ArrayBuffer[Tuple], Boolean)]()
+
+  override def processTexeraTuple(
+      tuple: Either[Tuple, InputExhausted],
+      input: Int,
+      pauseManager: PauseManager,
+      asyncRPCClient: AsyncRPCClient
+  ): Iterator[Tuple] = {
+    tuple match {
+      case Left(tuple) =>
+        if (input == 0) {
+          // left input, find on the right
+          val key = tuple.getField(buildAttributeName).asInstanceOf[K]
+          val (matchedTuples, _) =
+            rightTuples.getOrElse(key, (new ArrayBuffer[Tuple](), false))
+          val returnIter = matchedTuples
+            .map(right => {
+              join(tuple, right)
+            })
+            .iterator
+          building(tuple, input)
+          returnIter
+        } else {
+          // right input, find on the left
+          val key = tuple.getField(probeAttributeName).asInstanceOf[K]
+          val (matchedTuples, _) =
+            leftTuples.getOrElse(key, (new ArrayBuffer[Tuple](), false))
+          val returnIter = matchedTuples
+            .map(left => {
+              join(left, tuple)
+            })
+            .iterator
+          building(tuple, input)
+          returnIter
+        }
+      case Right(_) =>
+        Iterator()
+    }
+  }
+
+  private def join(left: Tuple, right: Tuple): Tuple = {
+    val builder = Tuple
+      .newBuilder(operatorSchemaInfo.outputSchemas(0))
+      .add(left)
```

Comment:
Is there a case where the input left tuples and/or right tuples are already supporting retraction?